### PR TITLE
CUDA/HIP/SYCL preliminaries + misc improvements

### PR DESCRIPTION
**Pull request by FabioLuporini**
I want to use the JIT backdoor to play with CUDA/HIP/SYCL.

This PR goes in this direction adding misc features concerning both execution and code generation

### CODE REVIEW

File: devito/passes/iet/langbase.py [hunk 1]
```diff
@@ -43,6 +46,48 @@ class LangBB(object, metaclass=LangMeta):
...
 
+    @classmethod
+    def _map_data(cls, f):
+        if f.is_Array:
+            return f.symbolic_shape
+        else:
+            return tuple(f._C_get_field(FULL, d).size for d in f.dimensions)
+
+    @classmethod
+    def _make_symbolic_sections_from_imask(cls, f, imask):
```
Comment (FabioLuporini) [inline devito/passes/iet/langbase.py:57]: note: this is NOT new -- just lifted from `parpragma.py`

File: devito/passes/iet/langbase.py [hunk 2]
```diff
@@ -43,6 +46,48 @@ class LangBB(object, metaclass=LangMeta):
     Prodder = Prodder
     PointerCast = PointerCast
 
+    @classmethod
+    def _map_data(cls, f):
```
Comment (FabioLuporini) [inline devito/passes/iet/langbase.py:50]: note: this is NOT new -- just lifted from` parpragma.py`

File: devito/passes/iet/langbase.py [hunk 3]
```diff
@@ -298,3 +343,16 @@ def _(iet):
             return iet, {}
 
         return _initialize(iet)
+
+    def _is_offloadable(self, iet):
```
Comment (FabioLuporini) [inline devito/passes/iet/langbase.py:347]: note: this is NOT new -- just lifted from `parpragma.py`

File: devito/passes/iet/parpragma.py [hunk 4]
```diff
@@ -485,41 +474,6 @@ def _make_nested_partree(self, partree):
 
 class PragmaLangBB(LangBB):
 
-    @classmethod
-    def _make_symbolic_sections_from_imask(cls, f, imask):
```
Comment (FabioLuporini) [inline devito/passes/iet/parpragma.py:489]: see comments above... lifted to langbase.py

File: devito/types/basic.py [hunk 5]
```diff
@@ -1315,7 +1321,7 @@ class LocalObject(AbstractObject):
 # - To override SymPy caching behaviour
 
 
-class IndexedData(sympy.IndexedBase, Pickable):
+class IndexedData(sympy.IndexedBase, Pickable, CodeSymbol):
```
Comment (FabioLuporini) [inline devito/types/basic.py:1324]: the changes here are to pass a Function by "data pointer" rather by "struct dataobj *", ie it allows the generation of

```
void foo(float * f)
```

besides

```
void foo(struct dataobj * f_vec)
```

File: devito/arch/compiler.py [hunk 6]
```diff
@@ -597,8 +616,9 @@ def __lookup_cmds__(self):
     'pgcc': PGICompiler,
     'pgi': PGICompiler,
     'nvc': NvidiaCompiler,
-    'nvcc': NvidiaCompiler,
+    'nvc++': NvidiaCompiler,
     'nvidia': NvidiaCompiler,
+    'nvcc': CudaCompiler,
```
Comment (mloubout) [inline devito/arch/compiler.py:621]: would probably calll it `cuda` , nvc/nvc++ bit too close.

File: devito/ir/clusters/analysis.py [hunk 7]
```diff
@@ -135,7 +135,7 @@ def _callback(self, clusters, d, prefix):
         # so that `i`'s trip count can safely be rounded up
         # Note: autopadding guarantees that the padding size along the
         # Fastest Varying Dimension is a multiple of the SIMD vector length
-        functions = [f for f in scope.functions if f.is_Tensor]
+        functions = [f for f in scope.functions if f.is_AbstractFunction]
```
Comment (mloubout) [inline devito/ir/clusters/analysis.py:138]: So why this relaxation?

**Review by mloubout — COMMENTED** [review]
Looks good, minor comments. Is that callable now "nestable", i.e could you in theory add call to external functions?

**Review by georgebisbas — COMMENTED** [review]
Had a very quick look, nothing really caught my eye.  Question: Is the CUDA code gonna be linearized? Maybe better:
```
__global__
void saxpy(int n, float a, float *x, float *y)
{
  int i = blockIdx.x*blockDim.x + threadIdx.x;
  if (i < n) y[i] = a*x[i] + y[i];
}

```

Is thelinearization gonna help produce:
`int i = blockIdx.x*blockDim.x + threadIdx.x;` somehow?

File: devito/ir/clusters/analysis.py (hunk 7 above)
Comment (FabioLuporini) [reply devito/ir/clusters/analysis.py:138]: oh no, actually just `is_Tensor` being completely redundant, ie hitting same cases as `is_AbstractFunction` (which was more widespread and "sensible" as name)

**Comment by FabioLuporini** [comment]
@georgebisbas ultimately yes, but we're a long way from that :)

**Comment by FabioLuporini** [comment]
@mloubout 

> Is that callable now "nestable", i.e could you in theory add call to external functions?

what sort of external functions?

I guess not MPI routines for example :) they are external too... you mean your own C functions? if so:

* no, as in, Callables aren't really neastable; Calls are though (but that's not your point right?). 
* However, if that's what you up to, I got a prototypal branch in devitopro that adds an **API** to inject whatever custom call you like

File: devito/arch/compiler.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/arch/compiler.py:621]: that would overload `DEVITO_LANGUAGE=cuda`

I agree the names are close. But these are the NVidia names, really.

* nvcc is the NVidia CUDA Compiler
* nvc is the NVidia C compiler for C/C++/OpenACC/OpenMP/etc (but not CUDA)

**Comment by mloubout** [comment]
@FabioLuporini 
>what sort of external functions?

Was thinking stuff like blas/mkl/..... So yeah related to second point. I don't have a case for now just wondering.

File: devito/arch/compiler.py (hunk 6 above)
Comment (mloubout) [reply devito/arch/compiler.py:621]: I get you but I still find the name confusing, in particular with the redundant ones (`nvidia`, `nvc`) that make it even more confusing. 
Why is it an issue to have same name for language and arch?

**Review by georgebisbas — APPROVED** [review]

File: devito/arch/compiler.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/arch/compiler.py]: Not an issue. And I don't have strong objections. So changed it to `cuda` as per your suggestion

**Review by mloubout — APPROVED** [review]

**Pull request merged**